provider/cloudflare/cloudflare_test.go

```diff
@@ -3618,3 +3432,745 @@ func TestZoneService(t *testing.T) {
+		assert.False(t, ok, "Should not have custom hostname when none exist")
+	})
+}
+
+// TestApplyChangesWithCustomHostnames tests applying changes with custom hostnames
+func TestApplyChangesWithCustomHostnames(t *testing.T) {
+	ctx := context.Background()
+
+	t.Run("CreateWithCustomHostname", func(t *testing.T) {
+		client := NewMockCloudFlareClient()
+		provider := &CloudFlareProvider{
+			Client: client,
+			CustomHostnamesConfig: CustomHostnamesConfig{
+				Enabled: true,
+			},
+		}
+
+		changes := &plan.Changes{
+			Create: []*endpoint.Endpoint{
+				{
+					DNSName:    "origin.bar.com",
+					RecordType: endpoint.RecordTypeA,
+					Targets:    endpoint.Targets{"1.2.3.4"},
+					ProviderSpecific: endpoint.ProviderSpecific{
+						{
+							Name:  annotations.CloudflareCustomHostnameKey,
+							Value: "custom.example.com",
+						},
+					},
+				},
+			},
+		}
+
+		err := provider.ApplyChanges(ctx, changes)
+		assert.NoError(t, err)
+	})
+
+	t.Run("DeleteWithCustomHostname", func(t *testing.T) {
+		client := NewMockCloudFlareClient()
+		client.customHostnames = map[string][]CustomHostname{
+			"001": {
+				{
+					ID:                 "ch1",
+					Hostname:           "delete.example.com",
+					CustomOriginServer: "origin.example.com",
+				},
+			},
+		}
+		provider := &CloudFlareProvider{
+			Client: client,
+			CustomHostnamesConfig: CustomHostnamesConfig{
+				Enabled: true,
+			},
+		}
+
+		changes := &plan.Changes{
+			Delete: []*endpoint.Endpoint{
+				{
+					DNSName:    "origin.bar.com",
+					RecordType: endpoint.RecordTypeA,
+					Targets:    endpoint.Targets{"1.2.3.4"},
+					ProviderSpecific: endpoint.ProviderSpecific{
+						{
+							Name:  annotations.CloudflareCustomHostnameKey,
+							Value: "custom.example.com",
+						},
+					},
+				},
+			},
+		}
+
+		err := provider.ApplyChanges(ctx, changes)
+		assert.NoError(t, err)
+	})
+}
+
+// TestErrorHandling tests various error scenarios
+func TestErrorHandling(t *testing.T) {
+	ctx := context.Background()
+
+	t.Run("CustomHostname_ListError", func(t *testing.T) {
+		client := &mockCloudFlareClient{
+			Zones: map[string]string{"newerror-zone": "example.com"},
+		}
+		provider := &CloudFlareProvider{
+			Client: client,
+			CustomHostnamesConfig: CustomHostnamesConfig{
+				Enabled: true,
+			},
+		}
+
+		_, err := provider.listCustomHostnamesWithPagination(ctx, "newerror-zone")
+		assert.Error(t, err)
+	})
+
+	t.Run("CustomHostname_CreateError", func(t *testing.T) {
+		client := NewMockCloudFlareClient()
+		provider := &CloudFlareProvider{
+			Client: client,
+			CustomHostnamesConfig: CustomHostnamesConfig{
+				Enabled: true,
+			},
+		}
+
+		// Mock returns error for this specific hostname
+		ch := CustomHostname{
+			Hostname:           "newerror-create.foo.fancybar.com",
+			CustomOriginServer: "origin.example.com",
+		}
+
+		err := provider.Client.CreateCustomHostname(ctx, "001", ch)
+		assert.Error(t, err, "Should error for newerror-create hostname")
+	})
+}
```

Comment:
Don't improve coverage.
I'm not against a specific test for `convertCloudflareError()` but I'm not convinced by the way it's done.